### ISSUE TITLE
Unable to pass data from one pipeline to another

### ISSUE THREAD

**Issue by amajai** [intel/video-analytics-serving#66]
Hello,
I am planning to separate a VAS pipeline into two or more parts with different server ports. I would like to have each pipeline handle different process stages instead of having one single process. I used the VAS image to create 2 separate docker containers with its custom pipelines. So in order to pass the data from one pipeline to another, I decided to use gstreamer elements, `tcpclientsink` for sending data and `tcpserversrc` for receiving data. These are the elements used for each pipelines

VAS-Server A
```bash
{
  "type": "GStreamer",
  "template": ["urisourcebin name=source ! decodebin  ! queue ! tcpclientsink port=80802"],
  "description": "Pipeline A",
  "parameters": {
    "type": "object",
    "properties": {}
  }
}
```

VAS-Server B:
```bash
{
  "type": "GStreamer",
  "template": [
    "tcpserversrc port=80801 ",
    " ! gvadetect model={models[object_detection][1][network]} name=detection",
    " ! gvametaconvert name=metaconvert ! queue ! gvametapublish name=destination",
    " ! gvawatermark ! videoconvert ! appsink"
  ],
  "description": "Pipeline B",
  "parameters": {
    "type": "object",
    "properties": {...}
  }
}
```

I tried sending a POST request to Server A to see if the pipeline will work and transfer data over to Server B. But, I am met with this error:
```bash
{"levelname": "INFO", "asctime": "2021-03-12 08:06:19,926", "message": "Setting Pipeline 1 State to RUNNING", "module": "gstreamer_pipeline"}
{"levelname": "ERROR", "asctime": "2021-03-12 08:06:19,928", "message": "Error on Pipeline 1: gst-resource-error-quark: Error while sending data to \"localhost:80802\". (10): ../gst/tcp/gsttcpclientsink.c(218): gst_tcp_client_sink_render (): /GstPipeline:pipeline2/GstTCPClientSink:tcpclientsink1:\nOnly 234192 of 345600 bytes written: Error sending data: Connection reset by peer", "module": "gstreamer_pipeline"}
```

How can I resolve this? Thanks.

**Comment by nnshah1** [comment]
This is a very interesting use cases - but one we haven't had time to try.

Have you tried starting the listener first? I believe that is the order.

another option could be to make the first pipline a tcpserversink, something like:

```
gst-launch-1.0 filesrc location=/home/video-analytics-serving/samples/20200711_cat.mp4 ! decodebin name=decode decode. ! queue  ! tcpserversink host=10.54.71.15 port=8000 sync=true 
````

And then start the second one -

Either way - I suggest prototyping first with gst-launch, confirm the order (which side to start first), and then enable in va serving. 

Do you have a basic example based on gst-launch?

**Comment by amajai** [comment]
@nnshah1 Regarding the first attempt at transferring data, both the server and listener were both running.  Here are methods I have tried out so far for transferring data from one pipeline to another with specified host and port.

1. Same port/ from tcpserversink => tcpclientsrc

Error from Server A logs:
```bash
{"levelname": "ERROR", "asctime": "2021-03-14 09:59:20,820", "message": "Error on Pipeline 1: gst-resource-error-quark: Could not open resource for reading. (5): ../gst/tcp/gsttcpserversink.c(404): gst_tcp_server_sink_init_send (): /GstPipeline:pipeline2/GstTCPServerSink:tcpserversink1:\nFailed to bind on host 'localhost:8080': Error binding to address: Cannot assign requested address", "module": "gstreamer_pipeline"}
```
2. Same port/ from tcpclientsink => tcpserversrc


Error from Server A logs:
```bash
{"levelname": "INFO", "asctime": "2021-03-14 10:50:47,401", "message": "Creating Instance of Pipeline object-preprocess/1", "module": "pipeline_manager"}
{"levelname": "INFO", "asctime": "2021-03-14 10:50:49,307", "message": "Setting Pipeline 1 State to RUNNING", "module": "gstreamer_pipeline"}
{"levelname": "ERROR", "asctime": "2021-03-14 10:50:49,312", "message": "Error on Pipeline 1: gst-resource-error-quark: Error while sending data to \"localhost:8080\". (10): ../gst/tcp/gsttcpclientsink.c(218): gst_tcp_client_sink_render (): /GstPipeline:pipeline2/GstTCPClientSink:tcpclientsink1:\nOnly 234192 of 345600 bytes written: Error sending data: Connection reset by peer", "module": "gstreamer_pipeline"}
```

3. Different ports(8080 to 8081)/ from tcpclientsink => tcpserversrc

Error from Server A logs:

```bash
{"levelname": "INFO", "asctime": "2021-03-14 11:16:07,772", "message": "Creating Instance of Pipeline object-inference/1", "module": "pipeline_manager"}
{"levelname": "ERROR", "asctime": "2021-03-14 11:16:07,780", "message": "Error on Pipeline 1: gst-resource-error-quark: Could not open resource for reading. (5): ../gst/tcp/gsttcpserversrc.c(488): gst_tcp_server_src_start (): /GstPipeline:pipeline2/GstTCPServerSrc:tcpserversrc1:\nFailed to bind on host 'localhost:8080': Error binding to address: Cannot assign requested address", "module": "gstreamer_pipeline"}
```
4. Different ports(8080 to 8081)/ from tcpserversink => tcpclientsrc

Error from Server A logs:

```bash
{"levelname": "INFO", "asctime": "2021-03-14 11:24:26,413", "message": "Creating Instance of Pipeline object-preprocess/1", "module": "pipeline_manager"}
{"levelname": "ERROR", "asctime": "2021-03-14 11:24:26,416", "message": "Error on Pipeline 1: gst-resource-error-quark: Could not open resource for reading. (5): ../gst/tcp/gsttcpserversink.c(404): gst_tcp_server_sink_init_send (): /GstPipeline:pipeline2/GstTCPServerSink:tcpserversink1:\nFailed to bind on host 'localhost:8081': Error binding to address: Cannot assign requested address", "module": "gstreamer_pipeline"}
```

Doesn't seem like any of methods or orders work. However, as I was looking up solutions on how to split up a pipeline, I came across this [repo](https://github.com/matthew1000/gstreamer-cheat-sheet/blob/master/sharing_and_splitting_pipelines.md), where it talks about different methods for splitting and sharing up pipelines. It seems that to implement these methods, you have modify the python script to split and connect the pipelines. Is there a way to implement them within a pipeline template?

**Comment by nnshah1** [comment]
@amajai  One quick note: I believe you are trying to use 8080 - and that port is already used for the VA Serving rest interface. 

Can you try 8081 (both server sink and client)?

To see how to capture the output of one pipeline and inject into another (in the same process) you can look at:

https://github.com/intel/video-analytics-serving/blob/master/samples/app_source_destination/app_source_destination.py 

I still believe the best approach from a performance point of view would be to use native gstreamer interprocess channels - (shm / ipc ) - but in a pinch you can always create your own server / client pair by extending the app_source_destination sample

**Comment by amajai** [comment]
@nnshah1 Yes, I have both the elements connected to the same port from `tcpserversink` to `tcpclientsrc`. I finally managed to get pipeline running for Server A with no error. However, once the pipeline process ends, the data doesn't carry over to Server B. There wasn't any change within the Server B logs. Its like the data got lost after the `tcpserversink` process. Still haven't found a solution to resolve it or even a reason why it happens.

I might go with the option of following the link you provided, but it will take me some to understand how to utilize the process. Would much rather go with the template pipeline.

Do you know any good resource you could share on learning the native gstreamer interprocess channels?

Thanks.

**Comment by amajai** [comment]
Hello @nnshah1. Was wondering if its possible to use the elements `appsink` for sending data and `appsrc` for receiving data within the VAS pipelines. I have tried out using them by adding the docker container names of both VAS servers to the name parameter for the elements as a way to refer each other. These were the pipelines I used:

Server A
```bash

  "type": "GStreamer",
  "template": [
    "urisourcebin name=source ! decodebin ! queue ! appsink name=server-b"
  ],
```
 
Server B
```bash
  "type": "GStreamer",
  "template": [
    "appsrc name=server-a ! gvadetect model={models[object_detection][1][network]} name=detection",
    " ! gvametaconvert name=metaconvert ! queue ! gvametapublish name=destination",
    " ! gvawatermark ! videoconvert
  ],
```
Unfortunately I was not able to get it to work. So I wanted to confirm if the issue was from the name parameters or not using the `appsink` and `appsrc` elements correctly.

**Comment by nnshah1** [comment]
@amajai Sorry for the delay - we've been busy putting together the 0.5 release. Let me take a chance to look at this with fresh eyes and try to give some guidance next week. Please let me know any impending deadlines / milestones you are working with. Also if you can share the larger vision for distributed pipelines - that may help.
